specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,449 @@
+Then, it MUST calculate the voting power of all active delegations of the `_delegator`, and calculate the voting
+power adjustments for the new delegatee set using `_newDelegations`. Afterwards, the function MUST update
+the voting power of the current and new set of delegatees.
+
+The function MUST then sort the new delegatee set in descending order, and store the delegations as part of the
+`_delegations` mapping. While sorting, the function MUST check for the validity of the delegations, as specified
+in the [Delegation Validation](#delegation-validation) section. Additionally, the function MUST emit a
+`DelegateVotesChanged` event for each change in delegatee voting power.
+
+### Getters
+
+For backwards compatibility, the `GovernanceDelegation` contract MUST implement all public getter functions of the
+`GovernanceToken` related to delegation and voting power. These functions MUST be used by the
+`GovernanceToken` when an account has been migrated to the `GovernanceDelegation` contract. Otherwise,
+the `GovernanceToken` MUST use its own state. Similarly, all of the `GovernanceDelegation` getter functions
+MUST use the `GovernanceToken` state if the account has not been migrated.
+
+#### `checkpoints`
+
+Retrieves the checkpoints for a given user address.
+
+```solidity
+function checkpoints(address _account) external view returns (ERC20Votes.Checkpoint[] memory)
+```
+
+#### `numCheckpoints`
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `delegates`
+
+Retrieves the delegations of a given user address sorted in descending order by voting power.
+This function is intended to be used by the `GovernanceToken` contract to maximize for backwards compatibility.
+
+```solidity
+function delegates(address _account) external view returns (Delegation[] memory)
+```
+
+#### `getVotes`
+
+Retrieves the voting power of a given user address.
+
+```solidity
+function getVotes(address _account) external view returns (uint256)
+```
+
+#### `getPastVotes`
+
+Retrieves the voting power of a given user address at a block.
+
+```solidity
+function getPastVotes(address _account, uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `getPastTotalSupply`
+
+Retrieves the total supply of the `GovernanceToken` at a given block.
+
+```solidity
+function getPastTotalSupply(uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `migrated`
+
+Returns the migration status of an account — `True` if the account has been migrated, `False` otherwise.
+
+```solidity
+function migrated(address _account) public view returns (bool)
+```
+
+### Events
+
+#### `DelegationCreated`
+
+MUST trigger when an account delegates voting power to another address (delegatee).
+
+```solidity
+event DelegationCreated(address indexed account, Delegation delegation);
+```
+
+#### `DelegationsCreated`
```

Comment:
Worth asking data team if having a separate event is better or easier to just emit N events

Reply:
@chuxinh any thoughts on this? we could have just one event called `DelegationsCreated`

Reply:
updated to just use one event as they are very similar -- it will make it easier to index and for the implementation too.

Reply:
missed this comment. Are you referring to the above two events? I see the difference is that one is delegating to one address and the other is multiple ones. If there's no issue of combining them that should be fine to merge into one.

Also is this only created when they first delegate or every time?